core/src/main/java/com/graphhopper/routing/Router.java

```diff
@@ -102,111 +102,210 @@ public Router(GraphHopperStorage ghStorage, LocationIndex locationIndex,
+                if (result.paths.isEmpty())
+                    throw new RuntimeException("Empty paths for alternative route calculation not expected");
+
+                // each path represents a different alternative and we do the path merging for each of them
+                PathMerger pathMerger = createPathMerger(request, weighting, queryGraph);
+                for (Path path : result.paths) {
+                    ResponsePath responsePath = new ResponsePath();
+                    responsePath.setWaypoints(getWaypoints(qResults));
+                    pathMerger.doWork(responsePath, Collections.singletonList(path), encodingManager, translationMap.getWithFallBack(request.getLocale()));
+                    ghRsp.add(responsePath);
+                }
+                ghRsp.getHints().putObject("visited_nodes.sum", result.visitedNodes);
+                ghRsp.getHints().putObject("visited_nodes.average", (float) result.visitedNodes / (qResults.size() - 1));
+                return result.paths;
+            } else {
+                StopWatch sw = new StopWatch().start();
+                List<QueryResult> qResults = ViaRouting.lookup(encodingManager, request.getPoints(), weighting, locationIndex, request.getSnapPreventions(), request.getPointHints());
+                ghRsp.addDebugInfo("idLookup:" + sw.stop().getSeconds() + "s");
+                PathCalculator pathCalculator = createPathCalculator(qResults, profile, algoOpts, disableCH, disableLM);
+                // (base) query graph used to resolve headings, curbsides etc. this is not necessarily the same thing as
+                // the (possibly implementation specific) query graph used by PathCalculator
+                QueryGraph queryGraph = QueryGraph.create(ghStorage, qResults);
+                ViaRouting.Result result = ViaRouting.calcPaths(request.getPoints(), queryGraph, qResults, weighting.getFlagEncoder().getAccessEnc(), pathCalculator, request.getCurbsides(), forceCurbsides, request.getHeadings(), passThrough);
+
+                if (request.getPoints().size() != result.paths.size() + 1)
+                    throw new RuntimeException("There should be exactly one more point than paths. points:" + request.getPoints().size() + ", paths:" + result.paths.size());
+
+                // here each path represents one leg of the via-route and we merge them all together into one response path
+                ResponsePath responsePath = concatenatePaths(request, weighting, queryGraph, result.paths, getWaypoints(qResults));
+                responsePath.addDebugInfo(result.debug);
+                ghRsp.add(responsePath);
+                ghRsp.getHints().putObject("visited_nodes.sum", result.visitedNodes);
+                ghRsp.getHints().putObject("visited_nodes.average", (float) result.visitedNodes / (qResults.size() - 1));
```

Comment:
Returning some kind of `Result` type from `PathCalculator` makes a lot of sense to me yes. For `RoundTripRouting` and `ViaRouting` I am not so sure. Should they also the same `Result` class we would use in `PathCalculator` then? And how do you want to avoid these two lines this way? At some point we have to put the stuff from the result into the GHResponse. Or should we go back to manipulating the GHResponse inside ViaRouting etc.? I was trying to reduce side effects here, but not sure about this (and also this wasn't the focus here/there is definitely room for improvements architecture-wise here). Should we use some kind of generic Result type for such purposes (something like a `class Result<T> that can carry T=List<Path> plus a debug string and some hash map or something)?

Reply:
I tried using a result type for `PathCalculator` here: 5c7c22d655b29d7a3bc5fdf997437684c6d22c21, but I am not sure if this is really an improvement. WDYT?